Filter out local zones when selecting AZs in infra.go

annapoornanarayan opened the pull request:
For `Error: failed to wait for infrastructure stack creation: waiter state transitioned to Failure` on creating cluster due to IPv6 CIDR not being supported in local AZs.
*Issue #, if available:*

*Description of changes:*. 
Added filtering in `DescribeAvailabilityZones` to only request regular availability zones (`zone-type=availability-zone`). Modified `counts` to consider regular AZs only.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

wwvela commented:
* We might need to add filter for [DescribeAvailabilityZones here as well](https://github.com/aws/aws-k8s-tester/blob/main/internal/deployers/eksapi/infra.go#L97C2-L100), for case if there is no instance type passed.
* The IPv6 CIDR is generated by default here in [infra template](https://github.com/aws/aws-k8s-tester/blob/main/internal/deployers/eksapi/templates/infra.yaml#L87-L92), so we must pick AZs which support the IPv6 CIDR. Unless we add more flags to make it an optional. To quickly unblock the test now, filter local zone out is acceptable

wwvela commented on internal/deployers/eksapi/infra.go:
```diff
@@ -417,12 +417,34 @@ func (m *InfrastructureManager) getRankedAZsForInstanceTypes(opts *deployerOptio
 	if err != nil {
 		return nil, fmt.Errorf("failed to describe instance type offerings: %v", err)
 	}
+	// Get all regular AZs excluding Local Zones
+	azDetails, err := m.clients.EC2().DescribeAvailabilityZones(context.TODO(), &ec2.DescribeAvailabilityZonesInput{
+		Filters: []ec2types.Filter{
+			{
+				Name:   aws.String("zone-type"),
+				Values: []string{"availability-zone"},
+			},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe availability zones: %v", err)
+	}
```
* I feel this is kind of duplicate here https://github.com/aws/aws-k8s-tester/blob/main/internal/deployers/eksapi/infra.go#L97C2-L100. You can add the filter in above api call and pass it into `getRankedAZsForInstanceTypes` method
* And the [DescribeInstanceTypeOfferingsInput](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2#DescribeInstanceTypeOfferingsInput) allows you pass `location - The location. For a list of possible identifiers`, so you can pass the AZs in filter to get those AZs availble for instance type only

mselim00 replied:
just adding on to here, we could also perhaps add a new flag to control the "zone type" and make that default to availability-zone. that would allow us to still deploy in a local zone if we so chose, and I think it generally makes sense that we would want one zone type for the test resources

annapoornanarayan commented:
Summary of changes:
1. deployer.go:
- Added flag `ZoneType` in `deployerOptions `struct which defaults to `availability-zone` (`VerifyUpFlags()` ).
2. infra.go: 
- Filtered for `ZoneType` in `createInfrastructureStack` and added valid AZs to new list allowedAZs
- If initialization takes [InstanceType branch](https://github.com/aws/aws-k8s-tester/blob/386d5790bb62fb126066d1286fd0f176602fec7a/internal/deployers/eksapi/infra.go#L107) , it will now take allowedAZs as an argument into `getRankedAZsForInstanceTypes`.
- Within `getRankedAZsForInstanceTypes`, use allowedAZs to only get instances within valid AZs using filter on `location`.
- Also, revert `counts` to original structure
